Bugfix/make 'conan new' templates work for special chars in package name

domfellner opened the pull request:
Changelog: Bugfix: Avoid issues of ``conan new`` templates with packages containing special characters like ``-.+``
Docs: Omit

- Replace leftovers of 'package_name' in templates (Conan 1.x) with 'class_name', defaulting to patched 'name')
- Make 'conan new' templates work for special chars in package name ("+-._" are allowed)


- [ ] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

This is targeting develop2 and I didn't find corresponding docs, so I omitted them. Closed the former docs PR targeting develop

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=10468) <br/>All committers have signed the CLA.

memsharded commented on conans/assets/templates/new_v2_cmake.py:
```diff
@@ -204,21 +204,23 @@ def test(self):
 
 
-def get_cmake_lib_files(name, version, package_name="Pkg"):
+def get_cmake_lib_files(name, version, package_name="Pkg", symbol_name="pkg"):
```
The default ``symbol_name="pkg"`` is dangerous. It shouldn't have a default.

domfellner replied:
I didn't want to break the API of this function. Should I add it as a mandatory parameter before package_name (I wondered this has a default anyway)?

domfellner replied:
Thanks for the fast feedback anyway.

Sorry, I cannot find where the rendering context is defined in the develop2 branch.

Is this implemented yet? The template looks a bit WIP there, containing

```class {{class_name or name}}Recipe(ConanFile):```

and I cannot find the definition of `class_name` anywhere.

Maybe these variables should be renamed according to their formatting like upper_camel_case_name and lower_camel_case_name? This would be more verbose than object_name.

memsharded replied:
> I didn't want to break the API of this function. Should I add it as a mandatory parameter before package_name (I wondered this has a default anyway)?

It is purely internal function, probably called just once in code. No probe to change it.

memsharded replied:
> Is this implemented yet? The template looks a bit WIP there, containing

Yes, the ``conan new cmake_lib -d name=pkg -d version=0.1`` command is fully operational in alpha3 and in develop2 branch. 
The templates are jinja templates, ``-d class_name=Something`` will dynamically inject it.

domfellner replied:
Ah I see, I need to manually inject the definitions. It was filled automatically in Conan 1.x, but it makes sense to make this explicit.

memsharded replied:
Yes, we are moving the whole "new" functionality to pure jinja templates, and a thin command line layer.

domfellner replied:
OK have changed the commit. It's not really worth the effort any more 🤷 I'll close the corresponding docs PR

memsharded reviewed (changes requested):
This is modifying code that has already been ported to 2.0 (develop2) branch, under a new place, new cli syntax.

We'd prefer to focus the work there. Merging this will create too many conflicts while merging develop->develop2, and this PR is not a critical functionality, so if you could target this for 2.0 "develop2" branch that would be very appreciated.

domfellner committed:
Make 'conan new' templates work with special chars

If any of ['+', '.', '-'] is in given 'name', the cmake templates
resulted in corrupted C/C++ syntax.

This is fixed by replacing all occurrences with "_".

memsharded reviewed (commented):
I would like to ask, because I don't fully get the purpose of this. Is this just to have a prettier define, like ``HELLO_EXPORT`` instead of ``hello_EXPORT``? Because I am fine with doing ``{{name.upper()}}`` that would be simpler. Because the thing is that users learning the ``-d symbol_name=HELLO`` variable and purpose is unlikely. 

Recall that this ``conan new`` templates are purely for demoing, testing, and quickly starting things, by no mean it intends to be production C++ code, and it should be replaced anyway with real code.

domfellner commented:
> I would like to ask, because I don't fully get the purpose of this. Is this just to have a prettier define, like `HELLO_EXPORT` instead of `hello_EXPORT`? Because I am fine with doing `{{name.upper()}}` that would be simpler. Because the thing is that users learning the `-d symbol_name=HELLO` variable and purpose is unlikely.
> 
> Recall that this `conan new` templates are purely for demoing, testing, and quickly starting things, by no mean it intends to be production C++ code, and it should be replaced anyway with real code.

My initial intention was that I created a package "foo-bar", as our package name convention works with "-" as separator. This leads to `foo-bar_EXPORT` etc. which breaks the build. I regarded this as inconvenient and frustrating for new users and wanted to ensure the buildability . As "+-." are allowed in names, these characters needed to be removed from the variable before replacement in the template, but I fear this is only implementable with three `replace()` calls. Maybe that's the way to go?
I agree that this is rather a corner use case and this new variable is inconvenient as well.

As I wrote it actually happened here and if I was a Conan beginner, I would have been frustrated. But we can also just leave it as it is as most users will not come across this anyway.

memsharded commented:
> As I wrote it actually happened here and if I was a Conan beginner, I would have been frustrated. But we can also just leave it as it is as most users will not come across this anyway.

Now I completely get the source cause which I think you are right. It is important. And it is important to solve it in an automated way, because other users will be frustrated until they learn they have another ``-d something=xxx`` to fix it.

I think we can safely do a ``replace("-","_").upper()`` in the jinja template (yeah, no big problem if we need to do 3 replaces for +. too, it is more important the UX than the template being super pretty), so it will be solved automatically for everyone. What do you think?

domfellner committed:
Fix 'package_name' leftovers with 'class_name'

domfellner commented on conans/cli/api/helpers/new/cmake_exe.py:
```diff
@@ -13,7 +13,7 @@ class {{class_name or name}}Recipe(ConanFile):
     author = "<Put your name here> <And your email here>"
     url = "<Package recipe repository url here, for issues about the package>"
-    description = "<Description of {package_name} here>"
+    description = "<Description of {{ name }} package here>"
```
adapted this to cmake_lib.py

domfellner commented on conans/cli/api/helpers/new/cmake_lib.py:
```diff
@@ -60,20 +60,20 @@ def package_info(self):
 
 source_h = """#pragma once
-
+{% set define_name = name.replace("-", "_").replace("+", "_").replace(".", "_").upper() %}
```
This might serve as a hint that there is more to jinja than mere replacing, exposing the {% ... %} syntax

domfellner commented:
Okay I tried my luck. I'll pay for another round ;)

memsharded committed:
Merge branch 'develop2' into feature/add-symbol-name-to-new-templates

memsharded reviewed (approved):
Looking good, thanks!

memsharded commented:
Merged, thanks for contributing, will be released in next 2.0.0-alpha4
